src/Nethermind/Nethermind.Evm/Instructions/EvmInstructions.Stack.cs

```diff
@@ -103,30 +105,88 @@ public static void Push(int length, ref EvmStack stack, int programCounter, Read
+    public struct Op2 : IOpCount { public static int Count => 2; }
+
+    /// <summary>
+    /// Push operation for two bytes.
+    /// </summary>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static EvmExceptionType InstructionPush2<TTracingInstructions>(VirtualMachine vm, ref EvmStack stack, ref long gasAvailable, ref int programCounter)
+        where TTracingInstructions : struct, IFlag
     {
         const int Size = sizeof(ushort);
-        public static int Count => Size;
+        // Deduct a very low gas cost for the push operation.
+        gasAvailable -= GasCostOf.VeryLow;
+        // Retrieve the code segment containing immediate data.
+        ReadOnlySpan<byte> code = vm.EvmState.Env.CodeInfo.CodeSection.Span;
 
-        /// <summary>
-        /// Push operation for two bytes.
-        /// </summary>
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static void Push(int length, ref EvmStack stack, int programCounter, ReadOnlySpan<byte> code)
+        ref byte bytes = ref MemoryMarshal.GetReference(code);
+        int remainingCode = code.Length - programCounter;
+        Instruction nextInstruction;
+        if (!TTracingInstructions.IsActive &&
+            remainingCode > Size &&
+            ((nextInstruction = (Instruction)Add(ref bytes, programCounter + Size))
+                is Instruction.JUMP or Instruction.JUMPF))
         {
-            int usedFromCode = Math.Min(code.Length - programCounter, length);
-            if (usedFromCode == Size)
+            // If next instruction is a JUMP we can skip the PUSH+POP from stack
+            ushort destination = As<byte, ushort>(ref Add(ref bytes, programCounter));
+            if (BitConverter.IsLittleEndian)
             {
-                ref byte bytes = ref MemoryMarshal.GetReference(code);
-                // Optimized push for exactly two bytes.
-                stack.Push2Bytes(ref Unsafe.Add(ref bytes, programCounter));
+                destination = BinaryPrimitives.ReverseEndianness(destination);
+            }
+            UInt256 result = default;
+            AsRef(in result.u0) = destination;
+            if (nextInstruction == Instruction.JUMP)
+            {
+                // Deduct the mid gas cost for performing a jump.
+                gasAvailable -= GasCostOf.Mid;
             }
             else
             {
-                // Use default padding behavior.
-                stack.PushLeftPaddedBytes(code.Slice(programCounter, usedFromCode), length);
+                // Deduct the high gas cost for a conditional jump.
+                gasAvailable -= GasCostOf.High;
+                // Pop the condition as a byte reference.
+                ref byte condition = ref stack.PopBytesByRef();
+                if (IsNullRef(in condition)) goto StackUnderflow;
+                // If the condition is zero (i.e. false), don't perform the jump.
+                if (As<byte, Word>(ref condition) == default)
```

Comment:
This is effectively `InstructionJumpIf` reimplemented. As with `InstructionJump` it was easy, this makes me think whether it should call the actual `InstructionJumpIf` maybe? Would it inline nicely or not at all? In any case, at least extracting the const costs would be great.